Refactor TORCH_DISTRIBUTED_DEBUG implementation

cbalioglu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #73149
* #73167
* __->__ #73166
* #73144

This PR refactors, cleans up, and optimizes the implementation of `TORCH_DISTRIBUTED_DEBUG`. It also introduces three new user APIs: `get_debug_level()`, `set_debug_level()`, and `set_debug_level_from_env()` to retrieve and modify the debug level after a process has started.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

cbalioglu committed:
Refactor TORCH_DISTRIBUTED_DEBUG implementation

This PR refactors, cleans up, and optimizes the implementation of TORCH_DISTRIBUTED_DEBUG.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of TORCH_DISTRIBUTED_DEBUG.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of TORCH_DISTRIBUTED_DEBUG.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of TORCH_DISTRIBUTED_DEBUG.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of TORCH_DISTRIBUTED_DEBUG.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of TORCH_DISTRIBUTED_DEBUG.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

rohan-varma commented on torch/distributed/__init__.py:
```diff
@@ -65,3 +66,6 @@ def is_available() -> bool:
     from .remote_device import _remote_device
+
+
+torch._C._distributed_c10d._set_debug_level()
```
Does this mean that if I do something like

```
import torch.distributed as dist # runs __init__ that sets the debug level

os.environ["TORCH_DISTRIBUTED_DEBUG"] = "DETAIL"
init_process_group() # expect detail debugging here
```

would this use case still work? Basically a lot of internal use cases set the debug level at various points in python code (always before init process group/DDP) and expect the debug level set there to be respected. Would that still work after these changes (and can we possibly add a test if it is not covered)?

cbalioglu replied:
Good catch. Nope it won't work since it parses the environment variable during process initialization. This is a fairly peculiar use of an environment variable. Let me think what I can do to keep this PR BC-compatible.

rohan-varma reviewed (approved):
LGTM, stamping it but had one question as well

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of TORCH_DISTRIBUTED_DEBUG.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d_common.py:
```diff
@@ -698,20 +698,27 @@ def tearDown(self):
-        default_debug_mode = dist._get_debug_mode()
-        self.assertEqual(default_debug_mode, dist._DistributedDebugLevel.OFF)
+        default_debug_mode = dist.get_debug_level()
+        self.assertEqual(default_debug_mode, dist.DebugLevel.OFF)
```
can we also have a test where we call `dist.set_debug_level_from_env()` and no env is set, and ensure the debug level is still off without any issues?

cbalioglu replied:
Updated the test in the latest revision.

rohan-varma commented on torch/testing/_internal/distributed/distributed_test.py:
```diff
@@ -5092,7 +5092,7 @@ def test_ddp_logging_data_cpu(self):
                 return os.environ[var] if var in os.environ else "N/A"
 
-            os.environ["TORCH_DISTRIBUTED_DEBUG"] = "INFO"
+            dist.set_debug_level(dist.DebugLevel.INFO)
```
do we need a test for `dist.set_debug_level(dist.DebugLevel.DETAIL)` to ensure that works properly? I'm pretty sure there should be a test somwhere here that does `TORCH_DIST_DEBUG=DETAIL`, and ensures that a wrapper process group is created, we can have a similar test if you think that would be beneficial.

cbalioglu replied:
Yes that makes sense. I don't have much context about what assertions such test should include though. It would be great if someone more knowledgeable could work on that.

rohan-varma replied:
I can add that later

rohan-varma reviewed (approved):
LGTM, the new APIs make sense to me. Will follow up with the appropriate internal changes, and just some comments around testing

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of `TORCH_DISTRIBUTED_DEBUG`. It also introduces three new user APIs: `get_debug_level()`, `set_debug_level()`, and `set_debug_level_from_env()` to retrieve and modify the debug level after a process has started.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of `TORCH_DISTRIBUTED_DEBUG`. It also introduces three new user APIs: `get_debug_level()`, `set_debug_level()`, and `set_debug_level_from_env()` to retrieve and modify the debug level after a process has started.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

rohan-varma reviewed (approved):
LGTM, thanks for updating the docs as well!

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of `TORCH_DISTRIBUTED_DEBUG`. It also introduces three new user APIs: `get_debug_level()`, `set_debug_level()`, and `set_debug_level_from_env()` to retrieve and modify the debug level after a process has started.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]

cbalioglu committed:
Update on "Refactor TORCH_DISTRIBUTED_DEBUG implementation"


This PR refactors, cleans up, and optimizes the implementation of `TORCH_DISTRIBUTED_DEBUG`. It also introduces three new user APIs: `get_debug_level()`, `set_debug_level()`, and `set_debug_level_from_env()` to retrieve and modify the debug level after a process has started.

Differential Revision: [D34371226](https://our.internmc.facebook.com/intern/diff/D34371226/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D34371226/)!

[ghstack-poisoned]
